src/modules/flight_mode_manager/tasks/ManualAltitude/FlightTaskManualAltitude.cpp

```diff
@@ -273,20 +275,62 @@ void FlightTaskManualAltitude::_ekfResetHandlerHeading(float delta_psi)
 {
 	// Only reset the yaw setpoint when the heading is locked
 	if (PX4_ISFINITE(_yaw_setpoint)) {
-		_yaw_setpoint += delta_psi;
+		_yaw_setpoint = wrap_pi(_yaw_setpoint + delta_psi);
 	}
+
+	_yaw_error_lpf.reset(wrap_pi(_yaw_error_lpf.getState() + delta_psi));
+	_yaw_error_ref = wrap_pi(_yaw_error_ref + delta_psi);
 }
 
 void FlightTaskManualAltitude::_updateSetpoints()
 {
-	_stick_yaw.generateYawSetpoint(_yawspeed_setpoint, _yaw_setpoint, _sticks.getYawExpo(), _yaw,
-				       _is_yaw_good_for_control, _deltatime);
-	_acceleration_setpoint.xy() = _stick_tilt_xy.generateAccelerationSetpoints(_sticks.getPitchRoll(), _deltatime, _yaw,
+	float yaw_setpoint_uncorrected = wrap_pi(_yaw_setpoint - _yaw_correction);
+	const bool reset_required = _updateYawCorrection();
+
+	if (reset_required) {
+		_yaw_setpoint = NAN;
+		yaw_setpoint_uncorrected = NAN;
+	}
+
+	const float yaw_opposite_correction = wrap_pi(_yaw -
+					      _yaw_correction); // remove correction from the estimate in case it's used to generate the setpoint
+	_stick_yaw.generateYawSetpoint(_yawspeed_setpoint, yaw_setpoint_uncorrected, _sticks.getYawExpo(),
+				       yaw_opposite_correction,
+				       _deltatime);
+	_yaw_setpoint = wrap_pi(yaw_setpoint_uncorrected + _yaw_correction);
```

Comment:
Maybe stupid question to ask but: In Stabilized, Altitude and Position mode the only thing we want to achieve is to lock yaw when no turn rate is commanded. How about we just control the unaided state which just has the goal to not drift but doesn't represent a real-world heading to north. Wouldn't that amke everything easier since you don't have to adjust the setpoint on the fly in the right way but rather just have a possibility to lock yaw and the controller does the right thing. Here's some pseudo-code for the idea:

Client old:
`if (no_yaw_rate && !isfinite(_yaw_setpoint)) _yaw_setpoint = _yaw;`
Client new:
`_yaw_setpoint_locked = no_yaw_rate;`

Controller new:
`if (_yaw_setpoint_locked) yaw_error = unaided_heading - _unaided_heading_when_lock_started;`